include/midi_pedal.hpp

```diff
@@ -35,30 +43,46 @@ class MidiPedalAdapter {
 			return;
 		}
 
-		if (note < 64)
-			held_lo[channel] |= 1 << note;
-		else
-			held_hi[channel] |= 1 << (note - 64);
+		// is the note already singing? Retrigger
+		if (getSustained(note)) {
+			note_off_callback(channel, note);
+		}
 
+		// trigger the note
 		note_on_callback(channel, note, velocity);
+
+		// add note to held array
+		setHeld(note, 1);
+
+		if (pedal_down) {
+			// pedal is down so we want to sustain the note
+			setSustained(note, 1);
+		}
 	}
 
 	void note_off(uint8_t channel, uint8_t note) {
-		if (!pedal_down[channel]) {
-			if (note < 64)
-				held_lo[channel] &= ~(1 << note);
-			else
-				held_hi[channel] &= ~(1 << (note - 64));
 
+		// remove note from held array
+		setHeld(note, 0);
+
+		if (!pedal_down) {
 			note_off_callback(channel, note);
 		}
 	}
 
+	void setSustained(uint8_t note, bool value) { note_sustained[note] = value; }
+
+	bool getSustained(uint8_t note) { return (note_sustained[note]); }
+
+	void setHeld(uint8_t note, bool value) { note_held[note] = value; }
+
+	bool getHeld(uint8_t note) { return (note_held[note]); }
+
   private:
-	uint64_t held_lo[16];
-	uint64_t held_hi[16];
+	bool note_held[127];      // finger is pressing the note
```

Comment:
there are 128, not 127 midi notes. this array ranges from 0 to 126 only, leaving out the highest possible note.